sites/game-client/ui/main-ui.mjs

```diff
@@ -43,28 +64,90 @@ class MainUI {
    * 
    * @param target The element which wants add button, should be an Overlay or Modal.
    */
-  addToolbarButton(target)
+  addToolbarButton(target) {
+    // TODO(fanlan1210)
+  }
 
   /**
    * Remove an button from the Toolbar.
    * 
    * @param target The element which wants remove button, should be an Overlay or Modal.
    */
-  removeToolbarButton(target)
+  removeToolbarButton(target) {
+    // TODO(fanlan1210)
+  }
 
   // TODO: Stack visualization? Stack height limit?
   /**
    * Show notification in notification area.
    * @param msg The message want to be shown.
    * @param {Number} timeout The duraion of showing notification, the unit is millisecond.
    */
-  showNotification(msg, timeout);
+  showNotification(msg, timeout) {
+    // TODO(lisasasasa)
+  }
 
   // If we want to use individual overlay for meeting
   // position = 0...n
   // setTopbarOverlay(position, overlay);
   // count=0 for no topbar.
   // resizeTopbarOverlay(count);
+
+  // ========== Internal Methods ============
+  // Methods below should only be used by UI classes such as Model/Overlay.
+
+  /**
+   * Send a DOM element back into staging area.
+   */
+  restoreToStaging(dom) {
+    this.stagingDom.appendChild(dom);
+  }
+
+  /**
+   * set Overlay at the position.
+   * @param {OverlayPosition} position one of OverlayPosition.
+   */
+  setOverlay(position, overlay) {
+    // If there's anything in the target, clear it out.
+    if (this.overlays.has(position)) {
+      if (!this.overlays.get(position).hide()) return false;
+    }
+
+    this.overlays.set(position, overlay);
+    const container = this.overlayDom[position];
+    container.appendChild(overlay.dom);
+    return true;
+  }
+
+  /**
+   * Removes the overlay at position.
+   */
+  clearOverlay(position, overlay) {
+    if (!this.overlays.has(position)) {
+      // Already cleared.
+      return true;
+    }
+
+    if (this.overlays.get(position) !== overlay) {
+      // This is not the overlay.
+      console.assert(false, `Overlay at position ${position} mismatch, got ` +
+        `${this.overlays.get(position)} and ${overlay}`);
+      return false;
+    }
+    this.restoreToStaging(overlay.dom);
+    this.overlays.delete(position);
+    return true;
+  }
+
+  /**
+   * This is called whenever the window is resized.
+   * @private
+   */
+  _onResize(event) {
+    for (const overlay of this.overlays.values()) {
+      overlay.onResize(event);
```

Comment:
Maybe
a) generate a warning when an overlay does not implement `onResize()` in development environment
or
b) force every overlay to implement `onResize()`

This code is going to crash if an overlay does not implement `onResize()`.

Reply:
There's an empty Overlay.onResize(), so it won't crash. Derived classes may/can implement onResize() but it's not strictly required if doing nothing on onResize() event is intended. Comment is added/adjusted to reflect this fact.